outputmanager.cpp

```diff
@@ -11,16 +11,59 @@
 
 #include "outputmanager.h"
 
+#include <algorithm>
+#include <iterator>
+
 OutputManager::OutputManager(QObject *parent) : QObject(parent)
 {
 
 }
 
-void OutputManager::handleNewData(const QByteArray &data)
+void OutputManager::operator << (const QByteArray &data)
+{
+    // problem : if a buffer ending with "\r\n" happens to be cut between '\r' and '\n'
+    //           the 2 resulting QString will be interpreted as having 2 line jumps,
+    //           instead of 1, if displayed in a QTextEdit.
+    //           To avoid this, final '\r' are removed from the received buffer,
+    //           and prepended to the next one
+
+    // flag indicating that the previously received buffer ended with CR
+    static bool prev_ends_with_CR = false;
+
+    // append raw data to the buffer, untouched
+    _buffer.append(data);
+
+    QString to_send;
+    if (prev_ends_with_CR)
+    {
+        // CR was removed at the previous buffer, so now we prepend it
+        to_send.append('\r');
+        prev_ends_with_CR = false;
+    }
+
+    if (data.size() > 0)
+    {
+        QByteArray::const_iterator end_cit = data.cend();
+        if (data.endsWith('\r'))
+        {
+            // if buffer ends with CR, we don't copy it
+            end_cit--;
+            prev_ends_with_CR = true;
+        }
+        std::copy(data.begin(), end_cit, std::back_inserter(to_send));
+    }
+
+    // notify that we have new data
+    emit dataConverted(to_send);
+}
```

Comment:
Yes it s because i want to use the output manager as an helper class,
converting raw data into display able data.
The manipulation I'm doing is necessary to present the data,  it happens to
take place in a QTextEdit and that's why I explained this in the comments
but i guess the code would remain the same if I had to present the data in
a terminal for example.

For the removal of '\r',  I thought of it but if cutecom-ng had to display
the output of some program that use '\r' without '\n' (to print over  some
text already printed for example) then the result would be the same as
expected

manipulate      was be used with but didn't do it because   losing them
could potentially not have the desired
Le 30 juin 2015 08:33, "Gianni Valdambrini" <notifications@github.com> a
écrit :

> In outputmanager.cpp
> <https://github.com/develersrl/cutecom-ng/pull/6#discussion_r33543746>:
>
> > +
> > +    if (data.size() > 0)
> > +    {
> > +        QByteArray::const_iterator end_cit = data.cend();
> > +        if (data.endsWith('\r'))
> > +        {
> > +            // if buffer ends with CR, we don't copy it
> > +            end_cit--;
> > +            prev_ends_with_CR = true;
> > +        }
> > +        std::copy(data.begin(), end_cit, std::back_inserter(to_send));
> > +    }
> > +
> > +    // notify that we have new data
> > +    emit dataConverted(to_send);
> > +}
>
> C'è una ragione particolare per cui questo codice debba appartenere alla
> OutputManager? Ti faccio notare che se l'output manager contiene del codice
> che sa che successivamente (nella mainwindow) c'è un textedit si sta
> violando l'incapsulamento. Se c'è un problema con gli '\r' e il TextEdit mi
> viene da pensare che potrebbe risolverlo direttamente il MainWindow.
> In aggiunta, avevi provato a rimuovere del tutto gli '\r'?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/develersrl/cutecom-ng/pull/6/files#r33543746>.
>